### ISSUE TITLE
False positive with  vector reserve

### ISSUE THREAD

**Issue by AlexeyDmitriev** [mgehre/llvm-project#18]
So, code 
```
#include <vector>
int main()
{
    std::vector<int> v;
    v.reserve(2);
    v.push_back(1);
    auto it = v.begin();
    v.push_back(2);
    return *it;
}
```

gives a warning 
```
<source>:9:13: warning: passing a dangling pointer as argument [-Wlifetime]

    return *it;

            ^

<source>:8:5: note: modified here

    v.push_back(2);
```

while `v` is guaranteed not to reallocate.

I used compiler on godbolt.

clang version 7.0.0 (https://github.com/mgehre/clang 4385bf94e655affaf1555bfbde3ccbb5c16c3ee8) (https://github.com/llvm-mirror/llvm d8e175bca5919de13fe003cab6b10f23af3d055d)
